### ISSUE TITLE
bogus error message for unterminated heredoc terminator

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#12268]
Migrated from [rt.perl.org#114104](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=114104) (status was 'resolved')

Searchable as RT114104$

**Comment by p5pRT** [comment]
### From @mauke

#### Created by @mauke



\#\!perl
$\_ = \<\<"foo
foo
foo

\_\_END\_\_

Can't find string terminator "foo
" anywhere before EOF at foo\.pl line 2\.

The error message is bogus because there are in fact two "foo\\n"'s in the code\.
The real problem is that there is no second '"' to terminate the heredoc
terminator\. This error message should be improved\.

<details><summary>Perl Info</summary>

```

Flags:
    category=core
    severity=low

This perlbug was built using Perl 5.12.1 - Thu Jun  3 20:09:15 CEST 2010
It is being executed now by  Perl 5.16.0 - Mon May 21 12:24:16 CEST 2012.

Site configuration information for perl 5.16.0:

Configured by mauke at Mon May 21 12:24:16 CEST 2012.

Summary of my perl5 (revision 5 version 16 subversion 0) configuration:
   
  Platform:
    osname=linux, osvers=2.6.38-gentoo-r6, archname=i686-linux
    uname='linux nora 2.6.38-gentoo-r6 #1 preempt sat aug 6 03:05:34 cest 2011 i686 amd athlon(tm) 64 processor 3200+ authenticamd gnulinux '
    config_args='-Dcc=cgcc -Dprefix=/home/mauke/usr/local -Dman1dir=none -Dman3dir=none -Dinc_version_list=none -Doptimize=-O2 -flto'
    hint=recommended, useposix=true, d_sigaction=define
    useithreads=undef, usemultiplicity=undef
    useperlio=define, d_sfio=undef, uselargefiles=define, usesocks=undef
    use64bitint=undef, use64bitall=undef, uselongdouble=undef
    usemymalloc=n, bincompat5005=undef
  Compiler:
    cc='cgcc', ccflags ='-fno-strict-aliasing -pipe -fstack-protector -I/usr/local/include -D_LARGEFILE_SOURCE -D_FILE_OFFSET_BITS=64',
    optimize='-O2 -flto',
    cppflags='-fno-strict-aliasing -pipe -fstack-protector -I/usr/local/include'
    ccversion='', gccversion='4.6.3', gccosandvers=''
    intsize=4, longsize=4, ptrsize=4, doublesize=8, byteorder=1234
    d_longlong=define, longlongsize=8, d_longdbl=define, longdblsize=12
    ivtype='long', ivsize=4, nvtype='double', nvsize=8, Off_t='off_t', lseeksize=8
    alignbytes=4, prototype=define
  Linker and Libraries:
    ld='cgcc', ldflags ='-fstack-protector -L/usr/local/lib -O2 -flto'
    libpth=/usr/local/lib /lib/../lib /usr/lib/../lib /lib /usr/lib
    libs=-lnsl -lgdbm -ldb -ldl -lm -lcrypt -lutil -lc -lgdbm_compat
    perllibs=-lnsl -ldl -lm -lcrypt -lutil -lc
    libc=/lib/libc-2.14.1.so, so=so, useshrplib=false, libperl=libperl.a
    gnulibc_version='2.14.1'
  Dynamic Linking:
    dlsrc=dl_dlopen.xs, dlext=so, d_dlsymun=undef, ccdlflags='-Wl,-E'
    cccdlflags='-fPIC', lddlflags='-shared -O2 -flto -L/usr/local/lib -fstack-protector'

Locally applied patches:
    SAVEARGV0 - disable magic open in <ARGV>


@INC for perl 5.16.0:
    /home/mauke/usr/local/lib/perl5/site_perl/5.16.0/i686-linux
    /home/mauke/usr/local/lib/perl5/site_perl/5.16.0
    /home/mauke/usr/local/lib/perl5/5.16.0/i686-linux
    /home/mauke/usr/local/lib/perl5/5.16.0
    .


Environment for perl 5.16.0:
    HOME=/home/mauke
    LANG=en_US.UTF-8
    LANGUAGE (unset)
    LC_COLLATE=POSIX
    LD_LIBRARY_PATH=/home/mauke/usr/local/lib
    LOGDIR (unset)
    PATH=/home/mauke/usr/perlbrew/bin:/home/mauke/usr/local/bin:/usr/local/bin:/usr/bin:/bin:/opt/bin:/usr/i686-pc-linux-gnu/gcc-bin/4.4.5:/opt/sun-jdk-1.4.2.13/bin:/opt/sun-jdk-1.4.2.13/jre/bin:/opt/sun-jdk-1.4.2.13/jre/javaws:/opt/dmd/bin:/usr/games/bin
    PERLBREW_HOME=/home/mauke/.perlbrew
    PERLBREW_PATH=/home/mauke/usr/perlbrew/bin
    PERLBREW_ROOT=/home/mauke/usr/perlbrew
    PERLBREW_VERSION=0.27
    PERL_BADLANG (unset)
    PERL_UNICODE=SAL
    SHELL=/bin/bash

```
</details>

**Comment by p5pRT** [comment]
### From @jkeenan

On Thu Jul 12 03&#8203;:10&#8203;:38 2012\, l\.mai@&#8203;web\.de wrote&#8203;:
> 
> This is a bug report for perl from l\.mai@&#8203;web\.de\,
> generated with the help of perlbug 1\.39 running under perl 5\.16\.0\.
> 
> 
> \-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-
> \[Please describe your issue here\]
> 
> \#\!perl
> $\_ = \<\<"foo
> foo
> foo
> 
> \_\_END\_\_
> 
> Can't find string terminator "foo
> " anywhere before EOF at foo\.pl line 2\.
> 
> The error message is bogus because there are in fact two "foo\\n"'s in
>    the code\.
> The real problem is that there is no second '"' to terminate the
>    heredoc
> terminator\. This error message should be improved\.
> 

What error message would you like to see?

Thank you very much\.
Jim Keenan

**Comment by p5pRT** [comment]
The RT System itself - Status changed from 'new' to 'open'

**Comment by p5pRT** [comment]
### From @mauke

On 2012\-07\-13 James E Keenan via RT wrote&#8203;:

> On Thu Jul 12 03&#8203;:10&#8203;:38 2012\, l\.mai@&#8203;web\.de wrote&#8203;:
> > 
> > This is a bug report for perl from l\.mai@&#8203;web\.de\,
> > generated with the help of perlbug 1\.39 running under perl 5\.16\.0\.
> > 
> > 
> > \-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-
> > \[Please describe your issue here\]
> > 
> > \#\!perl
> > $\_ = \<\<"foo
> > foo
> > foo
> > 
> > \_\_END\_\_
> > 
> > Can't find string terminator "foo
> > " anywhere before EOF at foo\.pl line 2\.
> > 
> > The error message is bogus because there are in fact two "foo\\n"'s
> > in the code\.
> > The real problem is that there is no second '"' to terminate the
> >    heredoc
> > terminator\. This error message should be improved\.
> > 
> 
> What error message would you like to see?

"Can't find string terminator '"' anywhere before EOF at foo\.pl line 2\."

The construct that's not terminated is a "\.\.\." string\, not a
foo\-heredoc\. Same as for
\#\!perl
$\_ = "foo
foo
foo

\_\_END\_\_

Lukas

**Comment by p5pRT** [comment]
### From @cpansprout

On Thu Jul 12 03&#8203;:10&#8203;:38 2012\, l\.mai@&#8203;web\.de wrote&#8203;:
> 
> This is a bug report for perl from l\.mai@&#8203;web\.de\,
> generated with the help of perlbug 1\.39 running under perl 5\.16\.0\.
> 
> 
> \-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-
> \[Please describe your issue here\]
> 
> \#\!perl
> $\_ = \<\<"foo
> foo
> foo
> 
> \_\_END\_\_
> 
> Can't find string terminator "foo
> " anywhere before EOF at foo\.pl line 2\.
> 
> The error message is bogus because there are in fact two "foo\\n"'s in
>    the code\.
> The real problem is that there is no second '"' to terminate the
>    heredoc
> terminator\. This error message should be improved\.

There heredoc\-scanning code doesn’t care if the delimiter is missing\.

So it ends up using everything up to the end of the buffer as its
terminator \("foo\\n"\)\.  And then bug \#114102 kicks in\, preventing it from
seeing "\\nfoo\\n\\n"\.  The real bug here is that scan\_heredoc doesn’t
check that delimcpy actually found the final delimiter\.

With string eval\, I get to see the implicit ; that gets tacked on to the
end of the buffer&#8203;:

\#\!perl
eval'
$\_ = \<\<"foo
foo
foo

\_\_END\_\_
'
or die
\_\_END\_\_

This gives me&#8203;:

Can't find string terminator "
foo
foo
foo

\_\_END\_\_

;" anywhere before EOF at \(eval 1\) line 2\.
&nbsp;    \.\.\.propagated at \- line 2\.

\-\- 

Father Chrysostomos

**Comment by p5pRT** [comment]
### From @cpansprout

Fixed by 08b2a93\.

\-\- 

Father Chrysostomos

**Comment by p5pRT** [comment]
### From [Unknown Contact. See original ticket]

Fixed by 08b2a93\.

\-\- 

Father Chrysostomos

**Comment by p5pRT** [comment]
@cpansprout - Status changed from 'open' to 'resolved'
